helper/awsutil/region.go

```diff
@@ -0,0 +1,88 @@
+package awsutil
+
+import (
+	"fmt"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/session"
+	hclog "github.com/hashicorp/go-hclog"
+)
+
+// "us-east-1 is used because it's where AWS first provides support for new features,
+// is a widely used region, and is the most common one for some services like STS.
+const DefaultRegion = "us-east-1"
+
+var (
+	// Order here matters - if multiple values are set, the first one in this array wins.
+	RegionEnvKeys = []string{"AWS_REGION", "AWS_DEFAULT_REGION"}
+
+	ec2MetadataBaseURL = "http://169.254.169.254"
+)
+
+/*
+It's impossible to mimic "normal" AWS behavior here because it's not consistent
+or well-defined. For example, boto3, the Python SDK (which the aws cli uses),
+loads `~/.aws/config` by default and only reads the `AWS_DEFAULT_REGION` environment
+variable (and not `AWS_REGION`, while the golang SDK does _mostly_ the opposite -- it
+reads the region **only** from `AWS_REGION` and not at all `~/.aws/config`, **unless**
+the `AWS_SDK_LOAD_CONFIG` environment variable is set. So, we must define our own
+approach to walking AWS config and deciding what to use.
+
+Our chosen approach is:
+
+	"More specific takes precedence over less specific."
+
+1. User-provided configuration is the most explicit.
+2. Environment variables are potentially shared across many invocations and so they have less precedence.
+3. Configuration in `~/.aws/config` is shared across all invocations of a given user and so this has even less precedence.
+4. Configuration retrieved from the EC2 instance metadata service is shared by all invocations on a given machine, and so it has the lowest precedence.
+
+This approach should be used in future updates to this logic.
+*/
+func GetOrDefaultRegion(logger hclog.Logger, configuredRegion string) string {
+	if configuredRegion != "" {
+		return configuredRegion
+	}
+
+	for _, envKey := range RegionEnvKeys {
+		envVal := os.Getenv(envKey)
+		if envVal != "" {
+			return envVal
+		}
+	}
+
+	sess, err := session.NewSessionWithOptions(session.Options{
+		SharedConfigState: session.SharedConfigEnable,
+	})
```

Comment:
This will also pick up the region environment environment variables in the right order, so the lines just above this aren't necessary.

Reply:
Ah good point! OK, I stripped them out, re-ran the tests, and they verified the behavior is unchanged. Pushing that up and merging.